In [1]:
# Importing the Dataset
from sklearn import datasets
iris = datasets.load_iris()
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [2]:
import numpy as np
columns_names = ["Petal length","Petal Width","Sepal Length","Sepal Width"]
X = iris.data
Y = iris.target
target_names = iris.target_names
X,Y,target_names

(array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
        [5

In [3]:
# Now we have the data, lets write a recursive function, That will select a feature
# (Based on Gain Ratio). Since the data is continuous, We need to select a point in the 
# continuous data, according to which we will get the maximum Information 
import math

def entropy(X,Y,num_targets):
    ans = 0
    num_features = X.shape[1]
    if(Y.shape[0] == 0):
        return 0
    # Calculating entropy as sum(-pi log2 pi)
    total_points = Y.shape[0]
    for i in range(num_targets):
        if Y[Y==i].shape[0] > 0:
            ans -= ((Y[Y==i].shape[0] / total_points) * math.log2 ((Y[Y==i].shape[0] / total_points)))
    return ans

def DecisionTree(X,Y,features_used,feature_names,target_names,level = 0):
    # X -> Features
    # Y -> Target
    # features_used -> 1D array, which is used to specify which feature has been used, 0 is not used, 1 is used
    # feature_names -> Name of the features/columns
    # target_names -> Name of the target classes
    # level -> Current level of the node
    
    print("Level ", level)
    num_features = X.shape[1]
    count = []
    num_targets = len(target_names)
    # Let's get the count of all the target values
    for i in range(num_targets):
        count.append(Y[Y==i].shape[0])

    for i in range(len(target_names)):
        if(count[i] > 0):
            print("Count of ",i,"(",target_names[i],") "," = ", count[i],sep='')
    print("Current Entropy is =", entropy(X,Y,num_targets))
    
    if(entropy(X,Y,num_targets) == 0):
        # If Entropy = 0, It means leaf node
        print("Leaf Node")
    elif(features_used[features_used == 0].shape[0] == 0):
        # All the features are used to split and exhausted
        print("Leaf Node")
    else:
        feature_idx = -1
        max_gain_ratio = -1
        feature_split = -1
        # Go over all the features
        for i in range(num_features):
            if not features_used[i]:
                # Since the data is continuous, We have to check all the values, And take the maximum of all the values
                # As the values are like 3.4, 3.5, 3.6... We run the loop as 3.35, 3.45, 3.55...
                for j in np.arange(X[:,i].min() - 0.05,  X[:,i].max() + 0.05, 0.1):
                    # Splitting on ith index at value j
                    # Lets get the left array, and right array
                    X_left  = X[X[:,i] < j] 
                    X_right = X[X[:,i] > j]
                    Y_left  = Y[X[:,i] < j]
                    Y_right = Y[X[:,i] > j]
                                        
                    # Getting entropy of both the sides
                    entropy_left = entropy(X_left,Y_left,num_targets)
                    entropy_right = entropy(X_right,Y_right,num_targets)

                    size_total = Y.shape[0]
                    size_left = Y_left.shape[0]
                    size_right = Y_right.shape[0]
                    
                    # Calculating InfoGain
                    entropy_child = (entropy_left * size_left + entropy_right * size_right) / size_total
                    current_entropy = entropy(X,Y,num_targets)
                    info_gain = current_entropy - entropy_child
                    
                    # Calculating SplitInfo
                    split_info = 0
                    if size_left > 0:
                        split_info -= (size_left/size_total) * math.log2(size_left/size_total)
                    if size_right > 0:
                        split_info -= (size_right/size_total) * math.log2(size_right/size_total)

                    # print(feature_names[i], j, size_left, size_right, size_total, info_gain, split_info)

                    # Calculating Gain Ratio
                    if split_info > 0:
                        gain_ratio = info_gain/split_info
                    
                        if gain_ratio > max_gain_ratio:
                            feature_idx = i
                            feature_split = j
                            max_gain_ratio = gain_ratio
                    
        print('Splitting on feature',feature_names[feature_idx],'with gain ratio',round(max_gain_ratio, 3),"at value", round(feature_split, 3))
        features_used[feature_idx] = 1
        print()
        DecisionTree(X[X[:,feature_idx] < feature_split], Y[X[:,feature_idx] < feature_split], features_used,feature_names,target_names,level + 1)
        print()
        DecisionTree(X[X[:,feature_idx] > feature_split], Y[X[:,feature_idx] > feature_split], features_used,feature_names,target_names,level + 1)
        features_used[feature_idx] = 0

In [4]:
DecisionTree(X,Y,np.zeros(4),columns_names,target_names)

Level  0
Count of 0(setosa)  = 50
Count of 1(versicolor)  = 50
Count of 2(virginica)  = 50
Current Entropy is = 1.584962500721156
Splitting on feature Sepal Length with gain ratio 1.0 at value 1.95

Level  1
Count of 0(setosa)  = 50
Current Entropy is = 0.0
Leaf Node

Level  1
Count of 1(versicolor)  = 50
Count of 2(virginica)  = 50
Current Entropy is = 1.0
Splitting on feature Sepal Width with gain ratio 0.693 at value 1.75

Level  2
Count of 1(versicolor)  = 49
Count of 2(virginica)  = 5
Current Entropy is = 0.44506485705083865
Splitting on feature Petal length with gain ratio 0.498 at value 7.05

Level  3
Count of 1(versicolor)  = 49
Count of 2(virginica)  = 4
Current Entropy is = 0.3860189005698934
Splitting on feature Petal Width with gain ratio 0.063 at value 2.85

Level  4
Count of 1(versicolor)  = 27
Count of 2(virginica)  = 4
Current Entropy is = 0.5547781633412736
Leaf Node

Level  4
Count of 1(versicolor)  = 22
Current Entropy is = 0.0
Leaf Node

Level  3
Count of 2(virginic

### Now that we have a Decision Tree, It looks something like this

![Decision Tree](tree.png "Decision Tree")

In [5]:
class DecisionTreeNode:
    def __init__(self):
        self.level = -1
        self.count = [] # Count of different target values, At index 0 -> count of Setosa 
                        # At index 1 -> count of versicolor, At index 2 -> count of virginica
        self.isLeaf = False # Is the current node a leaf ?
        self.entropy = -1 # Entropy of current node
        self.splitFeature = -1 # Index of feature to split on
        self.splitValue = -1 # The value of the feature to split on
        self.left = None # The pointer to the left sub decision tree
        self.right = None # The pointer to the right sub decision tree
        
class DecisionTreeClassifier:
    def __init__(self):
        self.root = None
        # This is the root of the decision tree
        
    def entropy(self,X,Y,num_targets):
        ans = 0
        if(Y.shape[0] == 0):
            return 0
        # Calculating entropy as sum(-pi log2 pi)
        total_points = Y.shape[0]
        for i in range(num_targets):
            if Y[Y==i].shape[0] > 0:
                ans -= ((Y[Y==i].shape[0] / total_points) * math.log2 ((Y[Y==i].shape[0] / total_points)))
        return ans
    
    def fit(self,X,Y):
        features_used = np.zeros(X.shape[1]) # Array representing if the i'th feature, is used or not, 1 -> Used, 0 -> Not used
        level         = 0
        num_targets   = np.unique(Y).shape[0]
        self.root     = self.fitHelper(X,Y,features_used,level,num_targets)
        
    def fitHelper(self,X,Y,features_used,level,num_targets):
        # X -> Features
        # Y -> Target
        # features_used -> 1D array, which is used to specify which feature has been used, 0 is not used, 1 is used
        # level -> Current level of the node
        
        node = DecisionTreeNode()
        # We create a tree of a special node, Called decision tree node

        node.level = level
        
        # Let's get the count of all the target values
        for i in range(num_targets):
            node.count.append(Y[Y==i].shape[0])
        
        # Setting the entropy of current node
        node.entropy = self.entropy(X,Y,num_targets)

        if(node.entropy == 0):
            # If Entropy = 0, It means leaf node
            node.isLeaf = True
            return node
        
        elif(features_used[features_used == 0].shape[0] == 0):
            # All the features are used to split and exhausted
            node.isLeaf = True
            return node
        
        else:
            feature_idx = -1
            max_gain_ratio = -1
            feature_split = -1
            # Go over all the features
            for i in range(features_used.shape[0]):
                if not features_used[i]:
                    # Since the data is continuous, We have to check all the values, And take the maximum of all the values
                    # As the values are like 3.4, 3.5, 3.6... We run the loop as 3.35, 3.45, 3.55...
                    for j in np.arange(X[:,i].min() - 0.05,  X[:,i].max() + 0.05, 0.1):
                        # Splitting on ith index at value j
                        # Lets get the left array, and right array
                        X_left  = X[X[:,i] < j] 
                        X_right = X[X[:,i] > j]
                        Y_left  = Y[X[:,i] < j]
                        Y_right = Y[X[:,i] > j]

                        # Getting entropy of both the sides
                        entropy_left = self.entropy(X_left,Y_left,num_targets)
                        entropy_right = self.entropy(X_right,Y_right,num_targets)

                        size_total = Y.shape[0]
                        size_left = Y_left.shape[0]
                        size_right = Y_right.shape[0]

                        # Calculating InfoGain
                        entropy_child = (entropy_left * size_left + entropy_right * size_right) / size_total
                        current_entropy = self.entropy(X,Y,num_targets)
                        info_gain = current_entropy - entropy_child

                        # Calculating SplitInfo
                        split_info = 0
                        if size_left > 0:
                            split_info -= (size_left/size_total) * math.log2(size_left/size_total)
                        if size_right > 0:
                            split_info -= (size_right/size_total) * math.log2(size_right/size_total)
                            
                        # Calculating Gain Ratio
                        if split_info > 0:
                            gain_ratio = info_gain/split_info

                            if gain_ratio > max_gain_ratio:
                                feature_idx = i
                                feature_split = j
                                max_gain_ratio = gain_ratio

            # Splitting the feature at feature_idx index
            node.splitFeature = feature_idx
            node.splitValue = feature_split

            # Since the feature is exhaused, We say that the feature is used
            features_used[feature_idx] = 1
            
            # Recursively build the left Decision Tree, and the right Decision tree
            node.left = self.fitHelper(X[X[:,feature_idx] < feature_split], Y[X[:,feature_idx] < feature_split], features_used,level + 1,num_targets)
            node.right = self.fitHelper(X[X[:,feature_idx] > feature_split], Y[X[:,feature_idx] > feature_split], features_used,level + 1,num_targets)
            
            # Setting the feature back to 0, Because we are backtracking now
            features_used[feature_idx] = 0
            
            # The tree is built, So we return the node
            return node
        
    def printTree(self):
        self.printTreeHelper(self.root)
        return
        
    def printTreeHelper(self, root):
        print("Level :",root.level)
        print("Count :",root.count) 
        print("isLeaf :",root.isLeaf)
        print("Entropy :",root.entropy)
        if(root.isLeaf):
            print()
            return
        print("Split Feature :",root.splitFeature)
        print("Split Value", root.splitValue)
        print()
        if(root.left):
            self.printTreeHelper(root.left)
        if(root.right):
            self.printTreeHelper(root.right)
        return
        
    def predict(self,X):
        # X is a 2d Array, with every row as one data, and colums as features
        # ans_y contains the output for the ith data input
        ans_y = []
        for x in X:
            # Predict the answer for every data input, and append it to ans_y
            ans_y.append(self.predictHelper(x, self.root))
        return np.array(ans_y)
    
    def predictHelper(self, X, root):
        if root.isLeaf:
            # if node is leaf, we take the majority vote, i.e. the maximum value
            return root.count.index(max(root.count))
        else:
            # If the node is not a leaf, We check the splitFeature, And check if the value of that 
            # feature is less than the splitValue And based on that,
            # we move to left subtree or right subtree
            if(X[root.splitFeature] < root.splitValue):
                return self.predictHelper(X, root.left)
            else:
                return self.predictHelper(X, root.right)

In [6]:
model = DecisionTreeClassifier()
feature_names = ["Petal length","Petal Width","Sepal Length","Sepal Width"]
model.fit(X,Y)
model.printTree()

Level : 0
Count : [50, 50, 50]
isLeaf : False
Entropy : 1.584962500721156
Split Feature : 2
Split Value 1.9500000000000008

Level : 1
Count : [50, 0, 0]
isLeaf : True
Entropy : 0.0

Level : 1
Count : [0, 50, 50]
isLeaf : False
Entropy : 1.0
Split Feature : 3
Split Value 1.7500000000000007

Level : 2
Count : [0, 49, 5]
isLeaf : False
Entropy : 0.44506485705083865
Split Feature : 0
Split Value 7.049999999999993

Level : 3
Count : [0, 49, 4]
isLeaf : False
Entropy : 0.3860189005698934
Split Feature : 1
Split Value 2.8499999999999988

Level : 4
Count : [0, 27, 4]
isLeaf : True
Entropy : 0.5547781633412736

Level : 4
Count : [0, 22, 0]
isLeaf : True
Entropy : 0.0

Level : 3
Count : [0, 0, 1]
isLeaf : True
Entropy : 0.0

Level : 2
Count : [0, 1, 45]
isLeaf : False
Entropy : 0.15109697051711368
Split Feature : 0
Split Value 5.949999999999998

Level : 3
Count : [0, 1, 6]
isLeaf : False
Entropy : 0.5916727785823275
Split Feature : 1
Split Value 3.0500000000000007

Level : 4
Count : [0, 0, 6]
is

In [9]:
y_predicted = model.predict(X)
y_predicted

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [10]:
def score(ypred, ytrue):
    # Predicting SCORE : 
    # Correct Predictions / Total Predictions
    correct = 0
    total = 0
    for i in range(len(ypred)):
        if(ypred[i] == ytrue[i]):
            correct += 1
        total += 1
    return correct/total

print(score(y_predicted, Y))

0.9733333333333334


### The Score is 0.97
(It could be overfitting the data as well)